File: Venmo/src/test/java/com/braintreepayments/api/venmo/VenmoRequestUnitTest.java
package com.braintreepayments.api.venmo;

import static junit.framework.TestCase.assertTrue;
import static org.junit.Assert.assertEquals;

import android.os.Parcel;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

import java.util.ArrayList;

@RunWith(RobolectricTestRunner.class)
public class VenmoRequestUnitTest {

    @Test
    public void parcelsCorrectly() {
        VenmoRequest request = new VenmoRequest(VenmoPaymentMethodUsage.MULTI_USE);
        request.setDisplayName("venmo-user");
        request.setShouldVault(true);
        request.setProfileId("profile-id");
        request.setCollectCustomerBillingAddress(true);
        request.setCollectCustomerShippingAddress(true);
        request.setSubTotalAmount("10.00");
        request.setTaxAmount("1.00");
        request.setDiscountAmount("2.00");
        request.setShippingAmount("1.00");
        request.setTotalAmount("10.00");
        request.setFinalAmount(true);

        ArrayList<VenmoLineItem> lineItems = new ArrayList<>();
        lineItems.add(new VenmoLineItem(VenmoLineItemKind.DEBIT, "An Item", 1, "10.00"));
        request.setLineItems(lineItems);

        Parcel parcel = Parcel.obtain();
        request.writeToParcel(parcel, 0);
        parcel.setDataPosition(0);
        VenmoRequest result = VenmoRequest.CREATOR.createFromParcel(parcel);

        assertEquals(VenmoPaymentMethodUsage.MULTI_USE, result.getPaymentMethodUsage());
        assertEquals("venmo-user", result.getDisplayName());
        assertTrue(result.getShouldVault());
        assertEquals("profile-id", result.getProfileId());
        assertTrue(result.getCollectCustomerBillingAddress());
        assertTrue(result.getCollectCustomerShippingAddress());
        assertEquals("10.00", result.getSubTotalAmount());
        assertEquals("1.00", result.getTaxAmount());
        assertEquals("2.00", result.getDiscountAmount());
        assertEquals("1.00", result.getShippingAmount());
        assertEquals("10.00", result.getTotalAmount());
        assertEquals(1, result.getLineItems().size());
        assertEquals("An Item", result.getLineItems().get(0).getName());
        assertTrue(result.isFinalAmount());
    }
}

File: Venmo/src/test/java/com/braintreepayments/api/venmo/VenmoRequestUnitTest.kt
package com.braintreepayments.api.venmo

import android.os.Parcel
import kotlinx.parcelize.parcelableCreator
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

@RunWith(RobolectricTestRunner::class)
class VenmoRequestUnitTest {

    @Test
    fun `parcels correctly`() {
        val lineItems = ArrayList<VenmoLineItem>()
        lineItems.add(VenmoLineItem(VenmoLineItemKind.DEBIT, "An Item", 1, "10.00"))

        val sut = VenmoRequest(
            paymentMethodUsage = VenmoPaymentMethodUsage.MULTI_USE,
            displayName = "venmo-user",
            shouldVault = true,
            profileId = "profile-id",
            collectCustomerBillingAddress = true,
            collectCustomerShippingAddress = true,
            subTotalAmount = "10.00",
            taxAmount = "1.00",
            discountAmount = "2.00",
            shippingAmount = "1.00",
            totalAmount = "10.00",
            isFinalAmount = true,
            lineItems = lineItems
        )

        val parcel = Parcel.obtain().apply {
            sut.writeToParcel(this, 0)
            setDataPosition(0)
        }
        val result = parcelableCreator<VenmoRequest>().createFromParcel(parcel)

        assertEquals(VenmoPaymentMethodUsage.MULTI_USE, result.paymentMethodUsage)
        assertEquals("venmo-user", result.displayName)
        assertTrue { result.shouldVault }
        assertEquals("profile-id", result.profileId)
        assertTrue(result.collectCustomerBillingAddress)
        assertTrue(result.collectCustomerShippingAddress)
        assertEquals("10.00", result.subTotalAmount)
        assertEquals("1.00", result.taxAmount)
        assertEquals("2.00", result.discountAmount)
        assertEquals("1.00", result.shippingAmount)
        assertEquals("10.00", result.totalAmount)
        assertEquals(1, result.lineItems?.size)
        assertEquals("An Item", result.lineItems?.get(0)?.name)
        assertTrue(result.isFinalAmount)
    }
}
